anvil/sample.py

```diff
@@ -25,6 +27,175 @@ def sample(loaded_model, action, target_length: int, n_large=20000) -> torch.Ten
     current proposal. A uniform random number, u, is drawn and if u <= A then
     the proposed state phi^j is accepted (and becomes phi^i for the next update)
 
+    Parameters
+    ----------
+    loaded_model: Module
+        loaded_model which is going to be used to generate sample states
+    action: Module
+        the action upon which the loaded_model was trained, used to calculate the
+        acceptance condition
+    batch_size: int
+        the number of states to generate from the loaded_model
+    state_i: torch.Tensor or None
+        the current state of the chain. None if this is the first batch
+    log_ratio_i: float or None
+        the ratio corresponding to state_i, to be used in the acceptance step
+    first: bool
+        Flag indicating whether this is the first batch or not
+
+    Returns
+    -------
+    phi[chain_indices, :]: torch.Tensor
+        chain of configurations generated by the MH algorithm
+    history: torch.BoolTensor
+        boolean tensor containing accept/reject history of chain
+    state: torch.Tensor
+        the current phi state, for continuity between batches
+    log_ratio: float
+        the ratio associated with state
+    """
+    with torch.no_grad():  # don't track gradients
+        z = torch.randn((batch_size, loaded_model.size_in))  # random z configurations
+        phi = loaded_model.inverse_map(z)  # map using trained loaded_model to phi
+        log_ptilde = loaded_model(phi)
+    history = torch.zeros(batch_size, dtype=torch.bool)  # accept/reject history
+    chain_indices = torch.zeros(batch_size, dtype=torch.long)
+
+    log_ratio = log_ptilde + action(phi)
+    if not isfinite(exp(float(min(log_ratio) - max(log_ratio)))):
+        raise ValueError("could run into nans")
+
+    if first:  # first batch
```

Comment:
surely first is unneccessary when log_ratio_i and state_i are None by default, it's like extra bookkeeping for the sake of it

```python
if state_i is None and log_ratio i is None:
...
elif state_i is not None and log_ratio_i is not None:
...
else:
    raise SomeKindofError("both need to either be None or have values")
```
I mean honestly passing the log ratio is kind of extra book keeping (I know it was me that originally put that) because it means one could potentially give a pair of values which don't match. Then it simplifies massively: the function takes `state_i` then `log_ratio_i` and `first` are redundant

so then we have

```python
    with torch.no_grad():  # don't track gradients
        z = torch.randn((batch_size, loaded_model.size_in))  # random z configurations
        phi = loaded_model.inverse_map(z)  # map using trained loaded_model to phi
        if state_i is not None:
            phi[0] = state_i
        log_ptilde = loaded_model(phi)
```

Reply:
Ah yeah, first is unnecessary. I think I thought it was clearer at one point but it's definitely better to only pass `state_i` around.

Reply:
also, state_i for the rest of this function is literally just state[j-1] and so I'm thinking we could throw a tonne of local variables away which would make the whole thing cleaner.

Finally I would get this function to return just the stack of states and the history (in that order)
then in later uses of the function we can just get initial state by indexing the final state of the returned stack

that way you won't have to have the horrible unpacking of function that you have at some point:
```python
c, d = return_abcd()[2:]
```

Reply:
I do agree with the neatness side of things, but is there not an efficiency argument for avoiding indexing arrays more times than necessary?

Reply:
well yeah there would be except we have to index twice to save the state anyway:

```python
import torch
import numpy as np

BIG_NUM = 100000
def test_1():
    a = torch.rand(BIG_NUM)
    b = torch.zeros(BIG_NUM)
    for i in range(1, BIG_NUM):
        if np.random.rand() > 0.5:
            b[i] = a[i]
        else:
            b[i] = a[i-1]
    return b

def test_2():
    a = torch.rand(BIG_NUM)
    last = a[0]
    b = torch.zeros(BIG_NUM)
    for i in range(1, BIG_NUM):
        if np.random.rand() > 0.5:
            b[i] = a[i]
            last = a[i]
        else:
            b[i] = last
    return b

In [1]: %timeit test_1()                                                       
765 ms ± 74.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [2]: %timeit test_2()                                                       
922 ms ± 8.89 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

in fact we see that the second is slower